### PR TITLE
feat: expose canister upgrade options in CLI

### PR DESCRIPTION
**Pull request by lwshang**
SDK-1779

# Description

`dfx canister install` and `dfx deploy` takes options `--skip-pre-upgrade` and `--wasm-memory-persistence`.

`dfx deploy --mode` now takes the same possible values as `dfx canister install --mode`: install, reinstall, upgrade, auto.

To maintain backward compatibility, a minor difference between the two commands remains.
If the `--mode` is not set, `dfx deploy` defaults to "auto", while `dfx canister install` defaults to "install".

# How Has This Been Tested?

e2e tests in install.bash and deploy.bash.

# Checklist:

- [x] The title of this PR complies with [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/).
- [x] I have edited the CHANGELOG accordingly.
- [x] I have made corresponding changes to the documentation.

### CODE REVIEW

File: src/dfx/src/commands/deploy.rs [hunk 1]
```diff
@@ -156,7 +146,8 @@ pub fn exec(env: &dyn Environment, opts: DeployOpts) -> DfxResult {
             bail!("The --mode=reinstall is only valid when deploying a single canister, because reinstallation destroys all data in the canister.");
         }
         (Some(_), _) => {
-            unreachable!("The only valid option for --mode is --mode=reinstall");
+            // unreachable!("The only valid option for --mode is --mode=reinstall");
```
Comment (sesi200) [inline src/dfx/src/commands/deploy.rs:149]:
```suggestion
```

File: src/dfx/src/util/clap/install_mode.rs [hunk 2]
```diff
@@ -0,0 +1,80 @@
+use anyhow::bail;
+use clap::Args;
+use ic_utils::interfaces::management_canister::builders::{
+    CanisterUpgradeOptions, InstallMode, WasmMemoryPersistence,
+};
+
+use crate::lib::error::DfxResult;
+
+/// CLI options for the mode of installing canister.
```
Comment (sesi200) [inline src/dfx/src/util/clap/install_mode.rs:9]:
```suggestion
/// CLI options for the mode of canister installation.
```
I like your formulation below better

File: src/dfx/src/util/clap/install_mode.rs [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+                wasm_memory_persistence: w,
+            }),
+        };
+        if canister_upgrade_options.is_some() {
+            if self.mode.as_deref() == Some("upgrade")
+                || self.mode.as_deref() == Some("auto")
+                || self.mode.is_none()
+            {
+                Ok(Some(InstallMode::Upgrade(canister_upgrade_options)))
+            } else {
```
Comment (sesi200) [inline src/dfx/src/util/clap/install_mode.rs:66]: `auto -> InstallMode::Upgrade` is an issue:

This test
```bash
@test "specify install mode" {
  dfx_start
  assert_command dfx deploy --mode auto --skip-pre-upgrade
}
```

fails with

```
Caused by: The replica returned a rejection error: reject code CanisterError, reject message Error from Canister bkyz2-fmaaa-aaaaa-qaaaq-cai: Attempted to execute a message, but the canister contains no Wasm module..
   Please install code to this canister before calling it.
```

**Review by sesi200 — CHANGES_REQUESTED** [review]

File: src/dfx/src/util/clap/install_mode.rs (hunk 3 above)
Comment (lwshang) [reply src/dfx/src/util/clap/install_mode.rs]: I fix the behavior in this case by introduced the `InstallModeHint` enum which makes the upgrade options considered to be hints in "auto" mode.

For example, if a canister is not installed before, `dfx deploy --mode auto --skip-pre-upgrade` will actually be in "install" mode in which the upgrade options don't take effect.

**Review by sesi200 — APPROVED** [review]

**Pull request merged**